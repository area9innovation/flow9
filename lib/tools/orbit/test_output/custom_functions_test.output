Flow compiler (3rd generation)

Processing 'orbit' on http server
done in 0.15s

tests/custom_functions_test.orb: tests/custom_functions_test.orb:29:15: Parse error
let numbers = [5, 10, 15, 20, 25];
              ^
cgihost->quitCode == 1
