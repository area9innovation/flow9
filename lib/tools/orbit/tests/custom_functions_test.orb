// Test custom functions with RunCore support

// Define a factorial function using RunCore operators
fn factorial(n) = n is (
  n => 1 if n <= 1;
  n => n * factorial(n - 1)
);

// Define a custom function that uses string operations
fn formatMessage(name, age) = "Hello, " + name + "! You are " + i2s(age) + " years old.";

// Define a function that manipulates arrays
fn sumHelper(arr, idx, acc) = 
    idx is (
      i => acc if i >= length(arr);
      i => sumHelper(arr, idx + 1, acc + index(arr, idx))
    );

fn sumArray(arr) = sumHelper(arr, 0, 0);

// Test the factorial function
let fact5 = factorial(5);
let fact10 = factorial(10);

// Test the string formatting function
let msg = formatMessage("Alice", 30);

// Test the array sum function
let numbers = [5, 10, 15, 20, 25];
let total = sumArray(numbers);

// Higher-order functions (simulated since we don't have map/filter built-in)
fn helper(arr, func, idx, result) = (
    idx is ( 
      i => result  if  i >= length(arr);
      i => (let newResult = result + [func(index(arr, idx))];
	  	  println("Here");
          helper(arr, func, idx + 1, newResult))
	)
);

fn applyToEach(arr, func) = helper(arr, func, 0, []);

// Double each number
fn double(x) = x * 2;
let doubled = applyToEach(numbers, double);

// Print the results
println("Factorial of 5: " + i2s(fact5));
println("Factorial of 10: " + i2s(fact10));
println("Formatted message: " + msg);
println("Sum of array: " + i2s(total));
println("Original array: " + prettyOrbit(numbers));
println("Doubled array: " + prettyOrbit(doubled))