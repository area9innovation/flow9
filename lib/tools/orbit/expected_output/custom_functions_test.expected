Factorial of 5: 120
Factorial of 10: 3628800
Formatted message: Hello, Alice! You are 30 years old.
Sum of array: 75
Original array: [5, 10, 15, 20, 25]
Doubled array: [10, 20, 30, 40, 50]
Result: 0
